Add cut / copy / paste support to automation & script editor

karwosts opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Often when I am changing automations, I'll need to move conditions around, nest them inside other conditions, etc. This gives an easy way to rearrange things without dropping into YAML and having to manually excise sections, try to paste it in the right place, get the tabs right, not miss any `-`, etc. 

Some design considerations:
 - This does not interact with the system clipboard, but rather just stores content locally in page memory. So currently these actions are limited to copying within the same automation; you cannot copy an action, leave the automation editor, and paste it into a different automation. 
 - Copy now seems to have some redundancy with the existing "Duplicate" action. I'm wondering if Duplicate should be removed in favor of Copy, since it is not as flexible. Though copy/paste is two clicks instead of one for duplicate. 

![copy-paste-test](https://github.com/home-assistant/frontend/assets/32912880/21fcc059-f959-4989-9079-1a158befb6b6)


## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
    https://community.home-assistant.io/t/wth-automations-ui-one-click-copy-yaml-one-click-paste-yaml/480726
    https://community.home-assistant.io/t/wth-cant-i-copy-and-paste-actions-in-an-automation/472557
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

karwosts committed:
Add cut/copy/paste support to automation editor

noxhirsch commented:
This is insanely useful! I (as a user - not a dev) would suggest du keep "Duplicate" and group it together with Cut/Copy with another divider above Duplicate.
I'd also suggest to add "Paste above/below" in the context menu since reordering is at least another two or three clicks.

bramkragten commented:
This is cool! :tada:

We initially had the idea to make this possible with drag and drop (like you can now do in re-order mode, but then drag from a block into another block) but ran into issues with the library we use and shadowDOM and have not finished it, should we look at that first?

bramkragten commented:
> This is insanely useful! I (as a user - not a dev) would suggest du keep "Duplicate" and group it together with Cut/Copy with another divider above Duplicate.

+1 for the grouping and divider

piitaya commented:
Love it !

I think we can add this feature without nested drag and drop. Also we should activate drag and drop by default because the 3 dots are hidden when using reorder mode and people might want to copy paste and drop and drop at the same time. 

Small proposal : add a dedicated paste icon next to the add button instead of paste entry in the add menu?

karwosts committed:
regroup items in menu

karwosts commented:
Added dividers. I also moved "Edit Id" in the trigger menu from below duplicate to above it so it fit with the top group. 

![image](https://github.com/home-assistant/frontend/assets/32912880/b8c365b1-3e29-4d84-a2ba-7d7268bd6da4)

karwosts committed:
dummy commit retrigger workflow

karwosts committed:
revert dummy commit

karwosts commented:
> Small proposal : add a dedicated paste icon next to the add button instead of paste entry in the add menu?

Any preference?

![image](https://github.com/home-assistant/frontend/assets/32912880/d6ca0438-707f-4897-9735-25e8dea4447d)
![image](https://github.com/home-assistant/frontend/assets/32912880/9b907819-f911-49e8-bc9f-a1f035023b98)
![image](https://github.com/home-assistant/frontend/assets/32912880/1d3e0447-e9d7-418b-a64c-530c343d839e)
![image](https://github.com/home-assistant/frontend/assets/32912880/9858f84a-3391-4128-877f-eea5ff8ab1b0)

bramkragten commented:
Would not make it an outlined button maybe, so it stands out more?

piitaya commented:
Maybe just a icon button

karwosts commented:
IMO Icon button didn't really look right next to the other button, they're different sizes and different styles.

Until there's a clear consensus I'll probably leave it as-is as a menu option for now.

noxhirsch commented:
Can you paste multiple times  after a single cut/copy or only once? If you can do it multiple times (which I think would be useful), then I would prefer the menu item anyway. I would only prefer the button if it appears temporarily after the cut/copy and disappears again after pasting (then my favorite would be "📋PASTE").

karwosts commented:
> Can you paste multiple times after a single cut/copy or only once?

multiple times, The clipboard is never cleared, only overwritten by new copy/cut.

frenck commented:
@Daniel-dev22 Let's keep the discussion on the PRs on the level of reviewing the contents of the PR. Thanks 👍 

../Frenck

bramkragten reviewed (approved):
I think this is a good start, we can improve and iterate on it. Thanks! :tada:
